### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by CryptoSFor**

### CODE REVIEW

File: wallet_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		{
+			name: "positive balance",
+			w:    &Wallet{balance: 1},
+			want: 1,
+		},
+		{
+			name: "negative balance",
+			w:    &Wallet{balance: -1},
+			want: -1,
+		},
```
Comment (xartishuk) [inline wallet_test.go:31]: add cases where balance is not a whole number: `4.2`, `1000000000.001`, `0.0000000000001`

File: wallet_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func TestWallet_Withdraw(t *testing.T) {
+	type args struct {
+		b Bitcoin
+	}
+	tests := []struct {
+		name    string
+		w       *Wallet
+		args    args
```
Comment (xartishuk) [inline wallet_test.go]: unnecessary complexity. You have 1 arg, just do `withdrawAmmount Bitcoin`, smth like that

File: wallet_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		wantErr bool
```
Comment (xartishuk) [inline wallet_test.go]: In our tests we want to not only check that an error occurred, but also that it was the exact error we expected to occur. To help with that, we should use the fact that "errors are values" in Go. Define concrete error values and check for them specifically. Try to figure this out, ping me if you have any problems

https://go.dev/blog/errors-are-values

File: wallet_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	}{
+		{
+			name:    "1",
```
Comment (xartishuk) [inline wallet_test.go]: use proper names

File: wallet_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+			w:       &Wallet{balance: 1},
+			args:    args{b: 2},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := tt.w.Withdraw(tt.args.b); (err != nil) != tt.wantErr {
+				t.Errorf("Wallet.Withdraw() error = %v, wantErr %v", err, tt.wantErr)
+			}
```
Comment (xartishuk) [inline wallet_test.go]: check balance after withdraw

File: wallet.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		return errors.New("negative input")
+	}
+
+	if w.balance < b {
+		return errors.New("insufficient funds in the wallet")
+	}
+
+	w.balance -= b
+
+	return nil
```
Comment (xartishuk) [inline wallet.go]: there is a bit of optimization you can do here. But we'll get to it later. Address other comments first

**Review by xartishuk — CHANGES_REQUESTED** [review]

File: wallet_test.go (hunk 1 above)
Comment (CryptoSFor) [reply wallet_test.go:31]: Done

File: wallet_test.go (hunk 2 above)
Comment (CryptoSFor) [reply wallet_test.go]: Done

File: wallet_test.go (hunk 1 above)
Comment (xartishuk) [reply wallet_test.go:31]: @CryptoSFor do that in other tests as well

File: wallet_test.go (hunk 2 above)
Comment (xartishuk) [reply wallet_test.go]: check the whole code

File: wallet.go (hunk 6 above)
Comment (xartishuk) [reply wallet.go]: @grigagod can you spot it?

File: wallet_test.go (hunk 1 above)
Comment (CryptoSFor) [reply wallet_test.go:31]: Done

File: wallet_test.go (hunk 2 above)
Comment (CryptoSFor) [reply wallet_test.go]: Done

File: wallet_test.go (hunk 4 above)
Comment (CryptoSFor) [reply wallet_test.go]: Done

File: wallet_test.go (hunk 5 above)
Comment (CryptoSFor) [reply wallet_test.go]: Done

File: wallet.go (hunk 6 above)
Comment (grigagod) [reply wallet.go]: of course @xartishuk, there is not need to lock wallet before checking for negative input and it's seems like a nice idea to move this checking into a separate method. Also if you are interested in optimization, take a look at [RWmutex](https://pkg.go.dev/sync#RWMutex)

File: wallet_test.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+			want:           0.01,
+			wantErr:        NegativeInputError,
+		},
+		{
+			name:           "Withdrawing from negative balance",
+			w:              &Wallet{balance: -0.001},
+			withdrawAmount: 0.1,
+			want:           -0.001,
+			wantErr:        InsufficientFundsError,
+		},
```
Comment (xartishuk) [inline wallet_test.go]: add a case where you have positive balance, but insufficient funds

File: wallet.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+	if float64(b) < 0 {
+		return NegativeInputError
+	}
+
+	if w.balance < b {
+		return InsufficientFundsError
+	}
+
+	w.balance -= b
```
Comment (xartishuk) [inline wallet.go:48]: if someone accidentally makes a slight change here:
```go
    w.balance = b
```
your tests will not fail. Can you find where the problem is?

**Review by xartishuk — CHANGES_REQUESTED** [review]
- add "testable example"
- move `mutex.Lock` per @grigagod suggestion

File: wallet_test.go (hunk 7 above)
Comment (CryptoSFor) [reply wallet_test.go]: Done

File: wallet.go (hunk 8 above)
Comment (CryptoSFor) [reply wallet.go:48]: Done

File: wallet/wallet.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		return InsufficientFundsError
+	}
+
+	w.balance -= b
+
+	return nil
+}
+
+// GetBalance allows getting balance of the wallet
+func (w *Wallet) GetBalance() Bitcoin {
```
Comment (xartishuk) [inline wallet/wallet.go]: oh, forgot one more thing: we don't use `Get` in getters:
https://golang.org/doc/effective_go#Getters

**Review by xartishuk — CHANGES_REQUESTED** [review]
For repos with a single package we don't usually use subfolders (`/wallet`). Move your go files back to the root of the repo. Use `git revert` to do that

File: wallet_test.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+				w.Withdraw(2)
+				defer wg.Done()
+			}(&wg)
+		}
+		wg.Wait()
+		var wantBalance Bitcoin = 20
+		if w.balance != wantBalance {
+			t.Errorf(" want = %v, got = %v", wantBalance, w.balance)
+		}
+	})
```
Comment (xartishuk) [inline wallet_test.go]: 1. Your deposits run and finish, only then your withdrawals run and finish. They should run simultaneously. 
2. Add some newlines

**Review by xartishuk — CHANGES_REQUESTED** [review]

File: wallet_test.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+		wg.Add(numOfDeposits + numOfWithdraws)
+
+		for i := 0; i < numOfDeposits; i++ {
+			go func(wg *sync.WaitGroup) {
+				w.Deposit(1)
+				defer wg.Done()
+			}(&wg)
+		}
+
+		for i := 0; i < numOfWithdraws; i++ {
```
Comment (xartishuk) [inline wallet_test.go]:
```suggestion
		numOfWithdrawals := 90

		wg.Add(numOfDeposits + numOfWithdrawals)

		for i := 0; i < numOfDeposits; i++ {
			go func(wg *sync.WaitGroup) {
				w.Deposit(1)
				defer wg.Done()
			}(&wg)
		}

		for i := 0; i < numOfWithdrawals; i++ {
```

**Review by xartishuk — APPROVED** [review]

**Pull request merged**